torchreid/engine/builder.py

```diff
@@ -1,63 +1,115 @@
-            enable_rsc=cfg.model.self_challenging_cfg.enable,
-            enable_sam=cfg.sam.enable,
-            should_freeze_aux_models=should_freeze_aux_models,
-            nncf_metainfo=nncf_metainfo,
-            initial_lr=initial_lr,
-            use_ema_decay=cfg.train.ema.enable,
-            ema_decay=cfg.train.ema.ema_decay
-        )
+        if not is_accuracy_aware_training:
+            engine = ImageAMSoftmaxEngine(
+                datamanager,
+                models=model,
+                optimizers=optimizer,
+                reg_cfg=cfg.reg,
+                metric_cfg=cfg.metric_losses,
+                schedulers=scheduler,
+                use_gpu=cfg.use_gpu,
+                save_chkpt=cfg.model.save_chkpt,
+                train_patience=cfg.train.train_patience,
+                early_stoping=cfg.train.early_stoping,
+                lr_decay_factor=cfg.train.lr_decay_factor,
+                conf_penalty=cfg.loss.softmax.conf_penalty,
+                label_smooth=cfg.loss.softmax.label_smooth,
+                aug_type=cfg.loss.softmax.augmentations.aug_type,
+                aug_prob=cfg.loss.softmax.augmentations.aug_prob,
+                decay_power=cfg.loss.softmax.augmentations.fmix.decay_power,
+                alpha=cfg.loss.softmax.augmentations.alpha,
+                size=(cfg.data.height, cfg.data.width),
+                max_soft=cfg.loss.softmax.augmentations.fmix.max_soft,
+                reformulate=cfg.loss.softmax.augmentations.fmix.reformulate,
+                pr_product=cfg.loss.softmax.pr_product,
+                softmax_type=softmax_type,
+                m=cfg.loss.softmax.m,
+                s=cfg.loss.softmax.s,
+                compute_s=cfg.loss.softmax.compute_s,
+                margin_type=cfg.loss.softmax.margin_type,
+                end_s=cfg.loss.softmax.end_s,
+                duration_s=cfg.loss.softmax.duration_s,
+                skip_steps_s=cfg.loss.softmax.skip_steps_s,
+                enable_masks=cfg.data.enable_masks,
+                adaptive_margins=cfg.loss.softmax.adaptive_margins,
+                class_weighting=cfg.loss.softmax.class_weighting,
+                attr_cfg=cfg.attr_loss,
+                base_num_classes=cfg.loss.softmax.base_num_classes,
+                symmetric_ce=cfg.loss.softmax.symmetric_ce,
+                mix_weight=cfg.mixing_loss.enable * cfg.mixing_loss.weight,
+                enable_rsc=cfg.model.self_challenging_cfg.enable,
+                enable_sam=cfg.sam.enable,
+                should_freeze_aux_models=should_freeze_aux_models,
+                nncf_metainfo=nncf_metainfo,
+                initial_lr=initial_lr,
+                use_ema_decay=cfg.train.ema.enable,
+                ema_decay=cfg.train.ema.ema_decay
+            )
+        else:
+            target_metric_name = 'top1'
+            engine = AccuracyAwareImageAMSoftmaxEngine(
+                datamanager,
```

Comment:
There are a lot of duplicate parameters here and in the original AMSoftmax and if we change something we should change it twice. Maybe combine these parameters in some way?

Reply:
Done